### PR TITLE
Helper function for defining differentiable solves.

### PR DESCRIPTION
**Pull request by shoyer**
``custom_implicit_solve`` is a helper function designed help library authors (most notably for ``jax.scipy.optimize``) define derivatives for functions that perform an implicit solve.

See the tests for an example of intended usage, defining a differentiable version of binary search.

This will start to become actually useful once we define some (differentiable) functions for matrix-free linear solves.

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
Brilliant!

I just realized (though we may have realized this before and I forgot...) an interesting comparison with scan.  In scan, the jvp rule produces a single scan where the inputs and outputs are augmented with tangent vectors as needed (analogous to just writing a solve like the ones here with a while_loop and then differentiating that, at least once we write the jvp rule of while_loop). Then to get reverse-mode, we do a partial eval of that scan jvp using a custom partial eval rule which unzips the computation into a primal part and a tangent part, each with a corresponding scan. The key point, though, is that the primal scan needs to save a bunch of linearization information from every iteration to pass of to the tangent scan, which (once that tangent computation is transposed) is used on the backward pass.

But the whole point here is that we don't need to save anything from the primal iterates: we're just going to do our tangent computation at the primal solution. And indeed when we partial eval this JVP rule we'll get exactly what we want: the primal solve and linearization of `func` at the solution is computed in the first computation, and the tangent solve is staged out into the second computation, saving only the data about the primal solution point that `vjp_func` needs.

Even better, this code expresses that the primal solve and tangent solve can be different routines in general (as opposed to the one particular approach of augmenting the primal iterations with tangent calculations, though we'd get that version if we just write the solve using `while_loop`, at least for forward-mode).

I'm interested to see if it makes sense to write our `fixed_point` routine from Autograd in terms of this `custom_implicit_solve` wrapper.

(One potential thorny issue here is when the Python functions involved, like `solve`, `tangent_solve`, and `func`, themselves close over traced arrays, e.g. for some outer differentiation trace. But that's exactly what `custom_transforms` solves for us! If we run into any issues on that front, it's an issue to be solved with `custom_transforms`, not in this code.)

**Comment by mattjj** [comment]
Actually, I need to revise what I said in that last parenthetical: I think `custom_transforms` will [check for and raise an error if it sees closed-over differentiated values](https://github.com/google/jax/blob/d9c938a48f7e4e0c4126abc074fb7fa53ac44a39/jax/api.py#L1309-L1314), but to handle them we need the JVP rule signature to change. That might take more thinking!

**Comment by mattjj** [comment]
With help from @dougalm, there are real issues with `func` having closed-over traces here, and that won't be caught anywhere. We might need to write this in terms of more core APIs to handle that  case. To handle differentiation of a higher-order function, we need to do stuff like in lax_control_flow.py. :/

Want to land and iterate, or brainstorm more?

File: jax/api.py
Comment (shoyer) [inline jax/api.py]: The math here isn't quite right.

Here's a derivation (will a comment in the next version):

      # Let F(u, m) = 0, where u = solution, m = params. Then:
      # d/dx F(u, m) = 0
      # ∂F/∂u du/dx + ∂F/∂m dm/dx = 0
      # du/dx = -(∂F/∂u)^{-1} ∂F/∂m dm/dx

I'm missing both the minus sign and `∂F/∂m`. This snuck through my test since the errors cancelled each other out.

**Comment by shoyer** [comment]
Well, we definitely shouldn't merge it yet since the math is wrong :).

Let me try to fix that first, then let's brainstorm how to get it fully working.

File: jax/api.py
Comment (shoyer) [reply jax/api.py]: Fixed

**Comment by shoyer** [comment]
Math should be fixed now if you want to take another look.

**Comment by mattjj** [comment]
Nice! I admit I didn't check the math before; usually I put my faith in numerical checks, but I guess the first time it got fooled.

If we put this in api.py, we should prefix this name with an underscore for now. Adding things to api.py is expensive, in that it defines the least-likely-to-break/disappear part of the API, and I realized later that this is going to take more work to make airtight. The reason is that we'll need to make some significant revisions to make differentiation work when `func` or `solve` close over traced values (e.g. values with respect to which we're differentiating). That's likely not a quick fix, and I'd have to think about how exactly to do it, but it'll probably end up looking something like what we do in lax_control_flow.py. (We should probably also add a warning to the docstring that this doesn't handle closed-over values in `func` or `solve`.)

Another option is to put this in `jax.experimental`.

Overall, though, I like the idea of landing this so we can iterate on it, since it already handles some useful cases and also defines a great API.

What do you think?


That was the main content of this follow-up review. This latter part is just a comment on math notation: I'd write the notation differently, because IMO "total derivatives" are just an obfuscation of the functions being differentiated. I'd also write the derivatives in terms of [positional arguments](https://mitpress.mit.edu/sites/default/files/titles/content/sicm_edition_2/preface001.html) rather than names, and avoid introducing `x` in addition to `u` and `m`. I'd write it like

```
F(u(m), m) = 0  # system of equations in m
∂_0 F(u(m), m) ∂ u(m) + ∂_1 F(u(m), m) = 0
∂ u(m) = - ∂_0 F(u*, m)^{-1} ∂_1 F(u*, m)
```

Now we have an expression for the Jacobian of the solution map `u(m)`, and so we can write down a VJP or a JVP pretty easily just by contracting on the left or right by a vector.

That's just a difference in notation though: the result, and the underlying meaning of the expressions, is the same.

**Comment by shoyer** [comment]
Here's how I'm inclined to write things down the math:
```
      # F(u(m), m) = 0  # system of equations in m
      # ∂/∂m F(u(m), m) = 0
      # ∂F/∂u ∂u/∂m + ∂F/∂m = 0
      # ∂u/∂m = -(∂F/∂u)^{-1} ∂F/∂m
```

I agree that it's better to write down a derivation for the Jacobian rather than the JVP.

It was not clear to me what the notation `∂ u(m)` means. It took me a while to figure out from context that it means the partial derivatives with respect to all arguments.

The notation `∂_0` also looks a little strange to me, though the main reason I don't like it is that it requires always explicitly requiring out `F(u, m)` which I think slightly obfuscates the (simple) math.

Also, if we use positional notation for partial derivatives, we should probably  rewrite `F(m, u(m))` or change the interface to expect `func(solution, params)`, because otherwise it will be confusing that the code uses the opposite indexing notation. Actually, maybe this worth making consistent anyways -- I am inclined to switch to `func(solution, params)` since `F(m, u(m))` looks a little funny to me.

**Comment by mattjj** [comment]
As for its provenance, that derivative notation is what we use in the Autodiff Cookbook, and it models JAX and Autograd's own differentiation API too. I got it from Spivak's Calculs on Manifolds, also used in Functional Differential Geometry and The Structure and Interpretation of Classical Mechanics; the latter is what I linked in my comment, though the Prologue section of [the free Functional Differential Geometry](https://mitpress.mit.edu/books/functional-differential-geometry) is probably even clearer in explaining why the notation is good. I find that notation to be the simplest, but certainly opinions on notation can differ, and I can't give a better defense than in SICM! I just wanted to offer an alternative. (By the way, something unfortunate about what you wrote is that ∂/∂m F(u(m), m) and ∂F/∂m mean different things! In what I wrote, there's only one thing going on: differentiation of functions.)

The notation isn't too important though, since we agree on the underlying math!

To clarify my comments about the code: there are several kinds of issues here that can crop up here when the functions involved close over values involved in differentiation. All the ones that `custom_transforms` can possibly catch lead to [explicit error messages](https://github.com/google/jax/blob/d9c938a48f7e4e0c4126abc074fb7fa53ac44a39/jax/api.py#L1309-L1314), but some can't be. In particular, if the only traced values are in the closure of `func`, there's no way for JAX to see them.

This closure problem involving higher-order primitives (i.e. primitives that take functions as arguments) is one we're familiar with, and one that we have two ways to solve: one is how we solve it with `jit` (which is a wrapper around the primitive `xla_call`, which takes a function argument) and the other is how we solve it in lax_control_flow.py. That latter approach also underlies how `custom_transforms` itself works. We'd probably want to use that technique again here.

**Comment by shoyer** [comment]
I'm happy to differ to you on notation, you've assuredly spent more time thinking about this than I have :)

This should be ready to merge now, assuming tests pass. Note that I swapped the expected order of the arguments to `func` to `func(solution, params)` to match the math.

**Comment by mattjj** [comment]
Excellent, thanks for this! Looking forward to following up on Thursday (and maybe on chat even sooner). In the meantime I'll try to land the no-more-tuples branch, which might make handling closures properly marginally easier.

**Pull request merged**
